MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm

```diff
@@ -24,9 +24,112 @@
 #include "MVKOSExtensions.h"
 #include "MVKFoundation.h"
 #include <sys/mman.h>
+#include <mach/mach_time.h>
+#include <inttypes.h>
 
 using namespace std;
 
+/** Converts a timestamp in GPU time into a timestamp in CPU time in nanoseconds, using the previously recorded timestamp correlation marker. */
+static MTLTimestamp adjustGPUTime(const TimestampCorrelationMarker& timestampCorrelationMarker, MTLTimestamp gpuTime)
+{
+	// Time in [0,1] domain = ( x - a ) / ( b - a )
+	double normalizedGpuTime = (gpuTime - timestampCorrelationMarker.gpuStart) / (double)(timestampCorrelationMarker.gpuEnd - timestampCorrelationMarker.gpuStart);
+	// Take the time in the [0,1] domain and scale it by the CPU
+	// lapse, then add the CPU start time to match the CPU’s timestamp.
+	double adjGpuTimeStamp = timestampCorrelationMarker.cpuStart + (timestampCorrelationMarker.cpuEnd - timestampCorrelationMarker.cpuStart) * normalizedGpuTime;
+	
+	// Convert from CPU counter to nanoseconds
+	mach_timebase_info_data_t time_info;
+	mach_timebase_info(&time_info);
+
+	return adjGpuTimeStamp * time_info.numer / time_info.denom;
+}
+
```

Comment:
We could probably avoid handling all this complexity ourselves by supporting `VK_EXT_calibrated_timestamps`.